components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java

```diff
@@ -492,4 +509,40 @@ private void doSend(Object key, ProducerRecord<Object, Object> record, KafkaProd
             kafkaProducer.send(record, cb);
         }
     }
+
+    private void startKafkaTransaction(Exchange exchange) {
+        exchange.getUnitOfWork().beginTransactedBy(transactionId);
+        kafkaProducer.beginTransaction();
+        exchange.getUnitOfWork().addSynchronization(new KafkaTransactionSynchronization(transactionId, kafkaProducer));
+    }
+}
+
+class KafkaTransactionSynchronization extends SynchronizationAdapter {
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaTransactionSynchronization.class);
+    private String transactionId;
+    private Producer kafkaProducer;
+
+    public KafkaTransactionSynchronization(String transacionId, Producer kafkaProducer) {
+        this.transactionId = transacionId;
+        this.kafkaProducer = kafkaProducer;
+    }
+
+    @Override
+    public void onDone(Exchange exchange) {
+        try {
+            if (exchange.getException() != null || exchange.isRollbackOnly()) {
```

Comment:
Good catch! We should handle such kafka exception, and in such cases, it don't need to call ```kafkaProducer.abortTransaction()```, is it right?

Reply:
I think ```exchange.getException()``` could catch the exception during the routing.

Reply:
It seems ```ProducerFencedException``` should be concerned
>This fatal exception indicates that another producer with the same <code>transactional.id</code> has been
 started. It is only possible to have one producer instance with a <code>transactional.id</code> at any
 given time, and the latest one to be started "fences" the previous instances so that they can no longer
 make transactional requests. When you encounter this exception, you must close the producer instance.

I'm not familiar with kafka. So @alexkazan87 can you check it?

Reply:
Correct! Generally, I would catch all the exceptions because "tomorrow" Kafka client could throw a different exception and throw it as a specific camel Kafka exception. So, you can handle and avoid the abort operation.

Reply:
I make some changes to abort a transaction only when the exception is not ```KafkaException```.